### PR TITLE
ahrs log active ekf

### PR DESCRIPTION
**Pull request by Jaaaky**
Allow AHRS to log active ekf type under AHR2, make it optional.
Activate this log for arduplane.

This is very useful when EKF type is changed during flight. I really change EKF from 2 to 3 for testing and it's not easy on logs to find which one was using at this specific time.
Also useful to know when fails to DCM for any reason.

### CODE REVIEW

**Comment by peterbarker** [comment]
Is there a good reason to put these changes into Plane only?

ahrs should not be being passed through - and nor should we add a new parameter based on a result from it :-)

**Comment by Jaaaky** [comment]
> Is there a good reason to put these changes into Plane only?

Actually, I could see that other vehicles use "always_use_ekf" and I don't know if they can "safely" switch ekf. But could easily be modified if approved.

> ahrs should not be being passed through - and nor should we add a new parameter based on a result from it :-)

I cannot get this point, if you can explain more?

**Comment by rmackay9** [comment]
Instead of adding the extra argument to Write_AHRS2 I think it might be possible to get the active ekf directly from the ahrs object

File: libraries/AP_Logger/LogStructure.h [hunk 1]
```diff
@@ -409,6 +409,7 @@ struct PACKED log_AHRS {
     int32_t lat;
     int32_t lng;
     float q1, q2, q3, q4;
+    int8_t ekf;
```
Comment (peterbarker) [inline libraries/AP_Logger/LogStructure.h]: Is this information available in the EKF logging itself?

If not, this should be a `uint8_t`....

Comment (Jaaaky) [reply libraries/AP_Logger/LogStructure.h]: No It's not available. The reason I've set it to int8_t is to set it to -1 if logging is not enabled.
But I'm rebasing now anyway.

**Comment by Jaaaky** [comment]
> Instead of adding the extra argument to Write_AHRS2 I think it might be possible to get the active ekf directly from the ahrs object.

Actually it was not available, I've added a getter. Please review.

> As a side note there's the active ekf type (DCM, EKF2, EKF3) and then also the active core (there's one core per IMU normally).

The active core is already logged as [NX]KF4.PI.

File: libraries/AP_Logger/LogStructure.h (hunk 1 above)
Comment (Jaaaky) [reply libraries/AP_Logger/LogStructure.h]: Rebased and almost re-written, please review.

**Comment by Jaaaky** [comment]
I've added a new variable called _active_ekf_type and allowed active_ekf_type() to change it.
I've done so to avoid calling the expensive active_ekf_type() on get_active_ekf_type() , just read its last change.
But I think this can cause a problem if the get_active_ekf_type() is called while active_ekf_type() so it may log intermediate value, "NONE" which it would actually be "EKF3".. I need your advice here.

**Comment by peterbarker** [comment]
@Jaaaky Eek.  Sorry, I really didn't think my comments through.  We actually bare another singleton, `AP_AHRS::ahrs_navekf()`, which gives you the sub-object.  It's pretty much idea for this sort of thing.  It's presence is gated behind a `#define`, so typical use for logging would be this:

```
#if AP_AHRS_NAVEKF_AVAILABLE
    const int8_t active_ekf_type = (int8_t)AP::ahrs_navekf().ekf_type();
#else 
    const int8_t active_ekf_type = -1;
#endif
```

**Comment by Jaaaky** [comment]
> @Jaaaky Eek. Sorry, I really didn't think my comments through. We actually bare another singleton, `AP_AHRS::ahrs_navekf()`, which gives you the sub-object. It's pretty much idea for this sort of thing. It's presence is gated behind a `#define`, so typical use for logging would be this:
> 
> ```
> #if AP_AHRS_NAVEKF_AVAILABLE
>     const int8_t active_ekf_type = (int8_t)AP::ahrs_navekf().ekf_type();
> #else 
>     const int8_t active_ekf_type = -1;
> #endif
> ```

Actually ekf_type is not active_ekf_type. ekf_type is what it should be used, active_ekf_type is what ahrs is actually using. If for example EKF2 fails to DCM due to no GPS/RelativeAID on plane/rover . active_ekf_type is 0 while ekf_type would be 2.

**Comment by Jaaaky** [comment]
The main idea for this PR is log what it's active. Maybe in the future some pr allows to fail from one EKF to another. Maybe the user change the AHRS_EKF_TYPE but actually it didn't switch for some reason.

After this I think I should also send another PR for EKF to log whether it's using RELATIVE, ABSOLUTE or NO AID.

This helps to debug scenarios where GPS/AIRSPEED/OPTFLOW/RNGFND are used together

**Comment by peterbarker** [comment]
On Sun, 27 Oct 2019, Jaaaky wrote:

> Actually ekf_type is not active_ekf_type. ekf_type is what it should be used, active_ekf_type is what ahrs is actually using. If for example EKF2 fails to

Yep, my goof.

It *was* a mock-up :-)

**Comment by Jaaaky** [comment]
@peterbarker So This was not merged, code has changed.
I'm really think it's useful. I've used it for few months now, and it could help me debug weird issues where I assumed that EKF was running(active) but actually it wasn't.
So how do you think this should be done, so I can re-do the code.

**Comment by peterbarker** [comment]
@Jaaaky I agree, knowing the actually active EKF type would be nice.

I don't think we need to add another accessor to get the active EKF type.  It's not that expensive and we don't run this that often.  Further, we already call it a *lot*, so fixing the expense should be done on the existing method rather than creating a new one (and wouldn't cause the cache coherency issue).

I think we need to think about where this belongs.  I'm not convinced by AHR2 - it's another, unrelated attitude (etc) solution, so I'm not sure why we would just throw it in there.  In fact, I think I might propose we change `Log_Write_AHR2` to `Log_Write_DCM`....  If we can wedge this information in somewhere then we know where to go to look for the currently-active estimator's logging, either the `XKF`/`NKF` stuff or the `DCM` message.  But it does mean this info has to live outside those messages....

**Comment by peterbarker** [comment]
The only suggestion I have as to where to put the currently active EKF is in the PM message.  That only gets logged every 10 seconds, so it won't catch all transitions.

I guess we could, instead, log active-ekf transitions rather than its current state?

**Comment by Jaaaky** [comment]
On transition it's nice, but where to log it?
For me, I think we should add new log AHRS message, that hold top level AHRS view of different states. Like 3d velocity, airspeed, active EKF, and so on.

**Comment by peterbarker** [comment]
> On transition it's nice, but where to log it?
> For me, I think we should add new log AHRS message, that hold top level AHRS view of different states. Like 3d velocity, airspeed, active EKF, and so on.

Possibly - I'd considered that, but hadn't thought about putting velocity in there.  People have previously proposed a `VEL` message to contain both angular and linar velocity.... We already have `POS` and `ATT` for example.

transition is easy - just log an `EV` or `ERR` with the code corresponding to the new AHRS in charge.

OTOH, if we did have a top-level AHRS message we could emit one of those on primary-estimator change, too.

**Comment by khancyr** [comment]
Done with  https://github.com/ArduPilot/ardupilot/pull/16166

**Pull request closed without merging**